### PR TITLE
Optimize SCE loss compute

### PR DESCRIPTION
**Pull request by pengwa**
### Optimize SCE loss compute

Compute optimization based on label data sparsity:
- Insert ShrunkenGather before SCELoss node, to filter out invalid labels for compute.
- Support ShrunkenGather upstream.
- Added test for the above.
- Added flag to enable label sparsity optimization with env var, by default disabled now. Will enable after comprehensive benchmarking later.
- Extract common logic into test_optimizer_utils.h/cc from core/optimizer/compute_optimzier_test.cc, then the common functions can be shared by both  core/optimizer/compute_optimzier_test.cc and  orttraining/core/optimizer/compute_optimzier_test.cc
- Extract common logic into shared_utils.h/cc: `GetONNXOpSetVersion` and `Create1DInitializerFromVector`

### CODE REVIEW

File: orttraining/orttraining/core/optimizer/compute_optimizer/sceloss_flops_reduction.cc [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+      // loss output is a scalar, don't need reset shape.
+    } else if (reduction.compare("none") == 0) {
+      need_reset_shape = true;
+    } else {
+      LOG_DEBUG_INFO(logger, "Skip SoftmaxCrossEntropyLossInternal node " + node.Name() +
+                                 " due to unsupported reduction type.");
+      continue;
+    }
+
+    if (need_reset_shape) {
```
Comment (baijumeswani) [inline orttraining/orttraining/core/optimizer/compute_optimizer/sceloss_flops_reduction.cc:126]: May we combine the if block with the `else if (reduction.compare("none") == 0)` block. We can change it in the future if we add support for it.

File: orttraining/orttraining/core/optimizer/compute_optimizer/sceloss_flops_reduction.h [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+ *                  SoftmaxCrossEntropyLossInternal(ignore_index=-100)
+ *                     /                       \
+ *   loss (shape is scalar or [token_count])      log_prob [token_count, classes]
+ *
+ * Be noted in Transformer-based models:
+ * > `token_count` usually equals with `batch size` x `sequence length`.
+ * > `classes` usually equals with `vocabulary`.
+ *
+ * Only insert ShrunkGather if all following conditions are met:
+ * 1. `SoftmaxCrossEntropyLossInternal`'s reduction MUST not be 'none', to make sure loss is a scalar.
```
Comment (baijumeswani) [inline orttraining/orttraining/core/optimizer/compute_optimizer/sceloss_flops_reduction.h:27]: Shall we use the logic in the *.cc file that checks what the reduction is istead of what it is not?

```suggestion
 * 1. `SoftmaxCrossEntropyLossInternal`'s reduction MUST be 'mean' or 'sum', to make sure loss is a scalar.
```

File: orttraining/orttraining/core/framework/ortmodule_graph_builder.cc [hunk 3]
```diff
@@ -152,7 +152,7 @@ Status OrtModuleGraphBuilder::OptimizeForwardGraph(std::unordered_set<std::strin
   Graph& forward_graph = forward_model_->MainGraph();
   ORT_RETURN_IF_ERROR(forward_graph.Resolve());
 
-  GraphTransformerManager graph_transformation_mgr{2};
+  GraphTransformerManager graph_transformation_mgr{5};
```
Comment (baijumeswani) [inline orttraining/orttraining/core/framework/ortmodule_graph_builder.cc:155]: What is the motivation behind this change?

File: cmake/onnxruntime_optimizer.cmake [hunk 4]
```diff
@@ -78,6 +78,8 @@ if (onnxruntime_ENABLE_TRAINING_APIS)
   list(APPEND onnxruntime_optimizer_src_patterns
     "${ORTTRAINING_SOURCE_DIR}/core/optimizer/*.h"
     "${ORTTRAINING_SOURCE_DIR}/core/optimizer/*.cc"
+    "${ORTTRAINING_SOURCE_DIR}/core/optimizer/compute_optimizer/*.h"
```
Comment (baijumeswani) [inline cmake/onnxruntime_optimizer.cmake:81]: thanks.

**Review by baijumeswani — APPROVED** [review]

**Review by baijumeswani — COMMENTED** [review]
Looks good.

**Comment by pengwa** [comment]
> Looks good.

Thank you, @baijumeswani
